cpp/src/arrow/util/utf8.h

```diff
@@ -350,6 +362,42 @@ static inline bool UTF8Decode(const uint8_t** data, uint32_t* codepoint) {
   return true;
 }
 
+static inline bool UTF8DecodeReverse(const uint8_t** data, uint32_t* codepoint) {
+  const uint8_t* str = *data;
+  if (*str < 0x80) {  // ascci
+    *codepoint = *str--;
+  } else if (ARROW_PREDICT_TRUE(Utf8IsContinuation(*str))) {
```

Comment:
Nit: instead of the `if` nesting here, it may be clearer to handle errors before:
```c++
if (ARROW_PREDICT_FALSE(!Utf8IsContinuation(*str))) {
  return false;
}
```